starlette/staticfiles.py

```diff
@@ -158,20 +157,25 @@ async def get_response(self, path: str, scope: Scope) -> Response:
         raise HTTPException(status_code=404)
 
     def lookup_path(
-        self, path: str
-    ) -> typing.Tuple[str, typing.Optional[os.stat_result]]:
+        self, path: Path
+    ) -> typing.Tuple[Path, typing.Optional[os.stat_result]]:
         for directory in self.all_directories:
-            full_path = os.path.realpath(os.path.join(directory, path))
-            directory = os.path.realpath(directory)
-            if os.path.commonprefix([full_path, directory]) != directory:
-                # Don't allow misbehaving clients to break out of the static files
-                # directory.
-                continue
+            original_path = Path(directory).joinpath(path)
+            full_path = original_path.resolve()
+            directory = Path(directory).resolve()
             try:
-                return full_path, os.stat(full_path)
+                stat_result = os.lstat(original_path)
+                full_path.relative_to(directory)
```

Comment:
We could also have used [`is_relative_to`](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to), instead of depending on `ValueError`.

Or just...
```python
            original_path = Path(directory).joinpath(path)
            full_path = original_path.resolve()
            directory = Path(directory).resolve()
            stat_result = os.lstat(original_path)
            if full_path.is_relative_to(directory):
                return full_path, stat_result
            elif full_path.is_symlink():
                return full_path, os.lstat(full_path)
```

Reply:
I also tried this before, the `is_relative_to` was introduced in Python 3.9, so couldn't use it.

Reply:
https://github.com/encode/starlette/pull/1377#discussion_r853867770